service/cert-reloader.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2023 LY Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: This code is based on athenz/k8s-athenz-sia's implementation:
+// TODO: https://github.com/AthenZ/k8s-athenz-sia/blob/main/pkg/util/cert-reloader.go
+// TODO: Yet, the original code is tailored specifically to k8s-athenz-sia's logic
+// TODO: So we could not copy the k8s-athenz-sia's cert-reloader code as is.
+// TODO: It would be beneficial to develop a more general-purpose library for this functionality in the future.
+// TODO: This way, both Garm and k8s-athenz-sia could utilize the same library.
+package service
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/kpango/glg"
+	"github.com/pkg/errors"
+)
+
+var defaultPollInterval = 1 * time.Minute
+
+// LogFn allows customized logging.
+type LogFn func(format string, args ...interface{})
+
+// CertReloader reloads the (key, cert) pair from the filesystem when
+// the cert file is updated.
+type CertReloader struct {
+	l            sync.RWMutex
+	certPath     string
+	keyPath      string
+	caPath       string
+	cert         *tls.Certificate
+	certPEM      []byte
+	keyPEM       []byte
+	caPool       *x509.CertPool // This is optional and can be nil
+	mtime        time.Time
+	pollInterval time.Duration
+	stop         chan struct{}
+}
+
+// GetCertFromCache returns the latest known certificate.
+func (w *CertReloader) GetCertFromCache() (*tls.Certificate, error) {
+	w.l.RLock()
+	c := w.cert
+	w.l.RUnlock()
+	return c, nil
+}
+
+// GetWebhook returns a function that is used to get X.509 Certificate stored in memory to connect to Athenz server
+// type IdentityAthenzX509 = func() (*tls.Config, error)
+func (w *CertReloader) GetWebhook() func() (*tls.Config, error) {
```

Comment:
rename?
```suggestion
func (w *CertReloader) GetTLSConfig() func() (*tls.Config, error) {
```

Reply:
This returns a function, not a config so no.
https://github.com/AthenZ/garm/blob/master/service/athenz_test.go#L166
Can be the following tho:
`GetAthenzWebhook()`
but yeah, not config.

Reply:
may be the following?
- `GetIdentityAthenzX509`
- `GetTLSConfigFunc`

Reply:
Will do `GetTLSConfigFunc()`